### DISCUSSION TITLE
Challenges creating a port for a library with non-library files dependencies

### DISCUSSION THREAD

**Discussion by asfernandes** [microsoft/vcpkg#35720] [General]
I'm trying to create a port for the Firebird database client library. Here is some characteristics of this library:
- Its main file (fbclient) should be a dynamic library
- It has dynamic libraries (say LoadLibrary) dependencies also built - stored in bin (Windows) or lib (Linux) directory
- It has dynamic libraries (say LoadLibrary) dependencies also built - stored in plugins directory
- It has non-library dependencies - I'm putting them in share/firebird/*
- All the libraries are relocatable in all platforms - the set of files can be put in any base directory and it will work correctly
- It's not a cmake project - but I created the scripts and the files are being copied as described below

To make discussion easy, I'll describe the Windows process.

The files are put in `vcpkg_installed/x64-windows`.

Here the first problem: `bin/fbclient.dll` will look for file `bin/../share/firebird/firebird.msg`. That is ok, but if the debug library is built with the same configure/patch way, it will not locate `share/firebird/firebird.msg` due to the extra level in `debug/bin` directory. `debug` and `share` are siblings, there is no `debug/share` in vcpkg conventions.

So I understand that the library should not be used (and actually I didn't want) directly in `vcpkg_installed/x64-windows`.

But when one makes a dependecy on `fbclient.dll`, only it is (and probably its link-time DLL dependencies) copied to the same directory as the target. That makes another problem as it not work without the other (dynamic libraries and non-library) files.

So I'm not understand how the port should be created to work with vcpkg/cmake.

The library is being built and since it's relocatable, it's a very easy to use library.

I do not think its characteristics are exclusive - many libraries depend on files.

**Comment by Osyotr** [comment]
One way to do it is to introduce `FIREBASE_DATA` environment variable that should be set by user (with possible default value `bin/../share/firebird`). Examples: `gdal` (`GDAL_DATA`), `proj` (`PROJ_DATA`), `osg` (`OSG_FILE_PATH`).

**Reply by asfernandes** [reply]
This is possible, using multiple environment variables.

But it's for special user configurations.

By default, it should work in the relocatable way.

**Reply by Osyotr** [reply]
On windows you first need to copy plugins and data into the output dir. There's a script that copies plugins for some libraries (https://github.com/microsoft/vcpkg/blob/master/scripts/buildsystems/msbuild/applocal.ps1), but there's no general mechanism yet (https://github.com/microsoft/vcpkg-tool/pull/974).
Personally I think it's user responsibility to copy this stuff into the output dir (i.e. users should find and copy it manually). After plugins/data are copied into the output dir, DLL that uses them can assume that they can be found in some predefined path in both debug and release builds of an application.
On linux you there's no requirement to copy dependencies into the output dir to be able to run application. The location of plugins/data then depends on whether it's debug build or release. I don't have answer for this except the one that I described in my previous post.

**Reply by asfernandes** [reply]
If it can't fully copy all necessary files, then how to avoid only the main library (and I guess, its linked DLL dependencies) to appear in the application bin directory that depends on the library?

**Reply by Osyotr** [reply]
https://github.com/microsoft/vcpkg/blob/0e47c1985273129e4d0ee52ff73bed9125555de8/scripts/buildsystems/vcpkg.cmake#L48

**Reply by asfernandes** [reply]
I set all this inside the port and in the app CMakeLists.txt, and the port's dll is stil copied to the app target location.

```
set(VCPKG_APPLOCAL_DEPS OFF)
set(X_VCPKG_APPLOCAL_DEPS_INSTALL OFF)
set(X_VCPKG_APPLOCAL_DEPS_SERIALIZED OFF)
```

What could be wrong?

**Reply by Osyotr** [reply]
These variables should be set before `vcpkg.cmake` is included, i.e. before first `project()` call or inside `CMakePresets.json` or passed to the command line. Moreover, only `VCPKG_APPLOCAL_DEPS` is enabled by default, so no need to set others. Also, these variables make no sense inside `portfile.cmake`.

**Reply by asfernandes** [reply]
Thanks. I was able to also set it with `option(VCPKG_APPLOCAL_DEPS OFF)` inside the main `CMakeLists.txt`.
